### PR TITLE
SIMD: Allow to switch off vectorized gather/scatter

### PR DESCRIPTION
**Pull request by kronbichler**
This is an attempt for a mitigation against the extremely annoying Intel GDS mitigation (aka Downfall): Add a flag to control the use of the `gather` and `scatter` instructions via `VectorizedArray`, which should shield our code base against the use of these instructions. Note that the performance tests https://dealii.org/performance_tests//reports/render.html?#!5824d3360e2ca5440695b4ac2da40146c42317a6.md discussed in #15873 document a performance degradation of our matrix-vector products by 20-40% (`timing-step_37/matvec_double` goes from 1.11s to 1.58s in the average case). As these are some of the highest-performing components, we need to take action **now**. The compiler might still insert `gather` instructions sometimes, because compilers will also try to use the opportunities and in fact those will only disappear once compilers have learned about the Downfall degradation, but this is out of my control. As I promised in https://github.com/dealii/dealii/issues/15873#issuecomment-1675903346, we might also do some changes on a larger scale.

@tamiko with this PR, you should be able to set `DEAL_II_USE_VECTORIZATION_GATHER=OFF`. I would be curious what the performance gets with this PR.

### CODE REVIEW

**Review by tjhei — APPROVED** [review]

**Comment by tjhei** [comment]
Can you add a changelog as well?

File: cmake/setup_cached_variables.cmake [hunk 1]
```diff
@@ -175,6 +176,12 @@ set(CMAKE_INSTALL_RPATH_USE_LINK_PATH "ON" CACHE BOOL
   )
 mark_as_advanced(CMAKE_INSTALL_RPATH_USE_LINK_PATH)
 
+option(DEAL_II_WITH_VECTORIZATION_GATHER
+  "For the x86 compilation target, the use of SIMD gather/scatter instructions can be much slower than using scalar loads. This includes a wide range of Intel hardware (in particular, server processors of the Broadwell, Skylake, Cascade Lake, and Ice Lake families released between 2015 and 2021). While the default is to aggressively use these instructions, this variable can be used to disable their use if deemed to give better performance."
+  ON
```
[the remark this answers is not in the document]
Comment (tamiko) [reply cmake/setup_cached_variables.cmake:181]: Not sure - this is really tricky. For example, my login node is unaffected but my compute nodes are very much affected.

**Review by tamiko — APPROVED** [review]

**Review by tamiko — CHANGES_REQUESTED** [review]

**Comment by tamiko** [comment]
@kronbichler I suggest we simply merge after the renaming. I have changed the configuration to set the toggle to false and we see what happens.

File: include/deal.II/base/config.h.in [hunk 2]
```diff
@@ -70,6 +70,7 @@
 #cmakedefine DEAL_II_WITH_TRILINOS
 #cmakedefine DEAL_II_WITH_UMFPACK
 #cmakedefine DEAL_II_FEATURE_UMFPACK_BUNDLED_CONFIGURED
+#cmakedefine DEAL_II_WITH_VECTORIZATION_GATHER
```
[the remark this answers is not in the document]
Comment (kronbichler) [reply include/deal.II/base/config.h.in]: Indeed, I changed the name. I am not entirely sure if the place in `config.h.in` is still right as this seems to be the only use of `_USE_`, but if we are unhappy we can always change it later.

**Comment by kronbichler** [comment]
> Can you add a changelog as well?

Changelog added, thank you for the remark.

**Review by tamiko — APPROVED** [review]

**Review by peterrum — APPROVED** [review]

File: cmake/setup_cached_variables.cmake (hunk 1 above)
[the remark this answers is not in the document]
Comment (tjhei) [reply cmake/setup_cached_variables.cmake:181]: Well, that is not something we can know. In situations like this you have to specify things manually.

**Comment by tamiko** [comment]
@kronbichler The results are split: https://dealii.org/performance_tests//reports/render.html?#!53ac485e6801587a97bce6ebca3cd11dc4f6ecb4.md

Most of the workloads improved - but did not fully recover. But, noticeable, with clang-16 some workloads regressed further.

**Pull request merged**